implemented support for composite primary keys and composite foreign keys...

k13gomez opened the pull request:
… for the following relationship types: has-one, belongs-to. issue #49.

I tried to make the solution as elegant yet simple as I could, includes support for has-one, has-many, belongs-to, and many-to-many.

I did not implement composite-keys support for "with-batch".

k13gomez commented on src/korma/core.clj:
```diff
@@ -487,11 +487,20 @@
+                                                  (filter (comp #{:belongs-to} :rel-type))
+                                                  (map :fk-key))
+                                      pk-key (-> query :ent :pk)
+                                      result-keys (concat
```
updated :dry-run exec-mode to handle seq pk-key and fk-key, to be able to generate dry-runs for composite key related tests.

k13gomez commented on src/korma/core.clj:
```diff
@@ -548,10 +557,16 @@
-  (let [fk-key (or fk (default-fk-name parent))]
-    {:pk (raw (eng/prefix parent (:pk parent)))
-     :fk (raw (eng/prefix child fk-key))
+(defn- to-raw-key [ent k]
```
new private helper function used to more easily implement composite-key support, but can be used elsewhere.

k13gomez commented on src/korma/core.clj:
```diff
@@ -548,10 +557,16 @@
-     :fk (raw (eng/prefix child fk-key))
+(defn- to-raw-key [ent k]
+  (if (seq? k)
+    (map #(raw (eng/prefix ent %)) k)
```
here the to-raw-key function handles convertion a seq of keys (composite pk-key or composite fk-key)

k13gomez commented on src/korma/core.clj:
```diff
@@ -548,10 +557,16 @@
+  (let [pk-key (or pk (:pk parent))
+        fk-key (or fk (default-fk-name parent))]
+    {:pk (to-raw-key parent pk-key)
+     :fk (to-raw-key child fk-key)
```
get-db-keys now uses to-raw-key to support composite-keys

k13gomez commented on src/korma/core.clj:
```diff
@@ -563,7 +578,7 @@
   (let [[db-keys foreign-ent] (db-keys-and-foreign-ent type ent sub-ent opts)
         fk-override (when-let [fk-key (:fk opts)]
-                      {:fk (raw (eng/prefix foreign-ent fk-key))
+                      {:fk (to-raw-key foreign-ent fk-key)
```
use to-raw-key in fk-override to support composite-keys

k13gomez commented on src/korma/core.clj:
```diff
@@ -652,8 +667,17 @@
-  (assoc ent :pk (keyword pk)))
+  ([ent pk]
+   (assoc ent :pk (keyword pk)))
+  ([ent pk & more]
```
overloaded pk function to set a composite-key by creating a pk seq.

k13gomez commented on src/korma/core.clj:
```diff
@@ -652,8 +667,17 @@
+  ([ent pk & more]
+   (assoc ent :pk (cons pk more))))
+
+(defn fk
```
new fk helper function to more easily set the :fk key to a key seq; this is optional but handy, see usage in new tests created.

k13gomez commented on src/korma/core.clj:
```diff
@@ -743,25 +767,36 @@
+                (select sub-ent
+                  (body-fn)
+                  (where
+                    (if (seq? ent-key)
```
implemented composite key support here by handling seq keys differently; white-space changes are also intentional to reduce line length and make the code more readable.

k13gomez commented on src/korma/core.clj:
```diff
@@ -743,25 +767,36 @@
                         table
-                        (= (raw (eng/prefix sub-ent sub-ent-key))
-                           (raw (eng/prefix (:ent query) ent-key))))))))
+                        (if (seq? ent-key) 
```
implemented composite key support here by handling seq-keys differently; this prevents existing tests from breaking by leaving handling for single pk/fk keys unmodified.

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -219,6 +219,26 @@
                   (with address)
                   (fields :address.state :name))))))
 
+(deftest with-has-one-composite-key
```
test composite-key handling with has-one relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -219,6 +219,26 @@
+          (select user-ck
+                  (with address-ck))))))
+
+(deftest with-belongs-to-composite-key
```
test composite-key handling with belongs-to relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -730,6 +750,14 @@
+(defentity state2-ck (transform identity) (pk :id :id2))
+(defentity address2-ck (belongs-to state2-ck (fk :sid :sid2)))
+
+(deftest test-belongs-to-with-transform-fn-for-subentity-composite-key
```
test composite-key handling with belongs-to relationship and transform fn for subentity

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -738,6 +766,14 @@
+(defentity address3-ck (transform identity))
+(defentity user3-ck (pk :id :id2) (has-one address3-ck (fk :uid :uid2)))
+
+(deftest test-has-one-with-transform-fn-for-subentity-composite-key
```
test composite-key handling with has-one relationship and transform fn for subentity

k13gomez commented:
I think I found a way to also implement one-to-many and many-to-many support for composite keys, I'm working on it at the moment, what I might leave out is with-batch support...

k13gomez commented on src/korma/core.clj:
```diff
@@ -303,10 +303,30 @@
-           (join* type ent (sfns/pred-= @(:rfk rel) (:rpk rel))))
-       (join* query type ent (sfns/pred-= (:pk rel) (:fk rel))))))
+       (let [{:keys [lpk lfk rpk rfk]} rel]
+         (-> query
```
support for composite-key on many-to-many joins

k13gomez commented on src/korma/core.clj:
```diff
@@ -303,10 +303,30 @@
+       (join* query type ent
+         (let [{:keys [pk fk]} rel]
+           (if (seq? pk)
+             (apply sfns/pred-and
```
support for composite-key on has-one and belongs-to joins

k13gomez commented on src/korma/core.clj:
```diff
@@ -541,17 +570,23 @@
+  {:lpk (to-raw-key parent (:pk parent))
+   :lfk (delay (to-raw-key {:table (name join-table)} @lfk))
+   :rfk (delay (to-raw-key {:table (name join-table)} @rfk))
+   :rpk (to-raw-key child (:pk child))
```
many-to-many now uses to-raw-key to support composite-keys

k13gomez commented on src/korma/core.clj:
```diff
@@ -541,17 +570,23 @@
    (var? ent) (recur @ent)
    :else      (throw (Exception. (str "Can't determine default fk for " ent)))))
 
+(defn- to-raw-key [ent k]
```
new internal function to-raw-key returns a single item for normal keys or a sequence for composite keys.

k13gomez commented on src/korma/core.clj:
```diff
@@ -720,7 +764,12 @@
-                                         (where {fk-key (get % pk)})))))))
+                                         (where (if (seq? pk)
+                                                  (apply hash-map
+                                                    (interleave
```
composite key support for one-to-many entity relationships

k13gomez commented on src/korma/core.clj:
```diff
@@ -743,36 +792,56 @@
+            (select ent
+              (join :inner join-table
+                    (if (seq? pk)
+                      (apply hash-map
```
composite key support for many to many

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -205,6 +205,17 @@
              (select user2
                      (with email)))))))
 
+(deftest with-has-many-composite-key
```
test composite-key handling with has-many relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -219,6 +230,54 @@
+          (select user-ck
+                  (with address-ck))))))
+
+(deftest join-has-one-composite-key
```
test composite-key handling for join has-one relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -219,6 +230,54 @@
+          (select user-ck
+                  (join address-ck))))))
+
+(deftest with-belongs-to-composite-key
```
test composite-key handling with belongs-to relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -219,6 +230,54 @@
+          (select address-ck
+                  (with user-ck))))))
+
+(deftest join-belongs-to-composite-key
```
test composite-key handling for join belongs-to relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -588,6 +647,16 @@
   (many-to-many mtm1 :mtm1_mtm2 {:lfk :mtm2_id
                                  :rfk :mtm1_id}))
 
+(defentity mtm3
```
many-to-many composite key enabled entities for testing

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -596,6 +665,15 @@
               "WHERE (\"mtm1_mtm2\".\"mtm2_id\" = ?) :: [1]\n")
          (with-out-str (dry-run (select mtm2 (with mtm1)))))))
 
+(deftest test-many-to-many-composite-key
```
test composite-key handling with many-to-many relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -604,6 +682,15 @@
               "WHERE (\"mtm1_mtm2\".\"mtm1_id\" = ?) :: [1]\n"))
       (with-out-str (dry-run (select mtm1 (with mtm2))))))
 
+(deftest test-many-to-many-reverse-composite-key
```
test composite-key handling with many-to-many reverse relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -612,6 +699,16 @@
               "ON \"mtm1_mtm2\".\"mtm1_id\" = \"mtm1\".\"id\" :: []\n")
          (with-out-str (dry-run (select mtm2 (join mtm1)))))))
 
+(deftest test-many-to-many-join-composite-key
```
test composite-key handling for join many-to-many relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -620,6 +717,16 @@
               "ON \"mtm1_mtm2\".\"mtm2_id\" = \"mtm2\".\"id\" :: []\n")
          (with-out-str (dry-run (select mtm1 (join mtm2)))))))
 
+(deftest test-many-to-many-join-reverse-composite-key
```
test composite-key handling for join many-to-many reverse relationship

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -730,6 +837,14 @@
+(defentity state2-ck (transform identity) (pk :id :id2))
+(defentity address2-ck (belongs-to state2-ck {:fk (fk :sid :sid2)}))
+
+(deftest test-belongs-to-with-transform-fn-for-subentity-composite-key
```
test composite-key handling for belongs-to relationship when transforms are enabled (with-one-to-one-later)

k13gomez commented on test/korma/test/core.clj:
```diff
@@ -738,6 +853,14 @@
+(defentity address3-ck (transform identity))
+(defentity user3-ck (pk :id :id2) (has-one address3-ck {:fk (fk :uid :uid2)}))
+
+(deftest test-has-one-with-transform-fn-for-subentity-composite-key
```
test composite-key handling for has-one relationship when transforms are enabled (with-one-to-one-later)

k13gomez commented:
I rebased the branch into 1 single commit so that it's bisect safe, it would be nice to include this in the next release.

immoh commented:
Thanks, I'll take a look at this when I have time.

immoh commented:
Ok, I finally had enough time to look through the PR. Here are my comments:

1. Wouldn't it make more sense to handle primary (and foreign) key always as a sequence of columns? Simple key would be a sequence of 1 column. This would simplify the rest of the code, you wouldn't have to always check if it's `seq?` or not.

2. I think it would be more clear to use `zipmap` instead of `(apply array-map (interleave ..))`. I guess the order doesn't really matter?

3. I am not happy with DSL for foreign key  `(has-many email-ck {:fk (fk :uid :uid2)})`. Maybe it could be simplified to `(has-many email-ck (fk :uid :uid2))`. I know it's a breaking change but that's ok.

4. Is with-batch still missing support for composite keys? What is the reason for this, is it difficult to implement?

No need to comment on individual diffs, it actually makes it harder for me to read the PR. I will ask if I have questions.

k13gomez commented:
Hi, here are my responses:

1. I initially did think about handling pk/fk as a sequence of columns but several tests "broke" because the way SQL is translated for pred-= is not wrapped in parenthesis like when the predicate is a map. I can probably work around this if I spend some more time though, or update the test expectations to reflect this change...i'll see which approach may be more feasible.

2. Using zipmap sounds cleaner, I'll look into refactoring it accordingly.

3. I setup the DSL this way so that it could be used when defining many-to-many relationships, since designing a DSL the way you suggested would only work for the fk and not for lfk, rfk. I'll think about it and see if I can come up with a workaround to improve it.

4. Yes, with-batch is missing support for composite keys because it relies on using an "in" predicate to find the related entities in a batched manner. This is not possible with composite keys because "in" only works with a single-key array. The only workaround I can think of would be to use a sequence of pred-= wrapped by a pred-or to obtain all the related entities in a batch. Let me know what you think, I suppose it would be better than no support at all.

Thanks for the tip on individual diff comments.

jg2356 commented:
I implemented changes according to the suggestions made, with some caveats to maintain 1.3, 1.4 compatibility (such as using azipmap instead of zipmap); also added with-batch support for composite keys. Take a look when you have time.

k13gomez commented:
I implemented changes according to the suggestions made, with some caveats to maintain 1.3, 1.4 compatibility (such as using azipmap instead of zipmap); also added with-batch support for composite keys. Take a look when you have time.

immoh commented:
I still think it would be better to have `pk` always create a sequence of keys, also for single key case. I don't mind if the generated sql changes in this case, we just need to modify the test cases.

I didn't get the compatibility issue - zipmap has been clojure.core from the beginning, hasn't it?

The current implementation using zipmap is too complicated:

```clojure
(apply sfns/pred-and
       (for [[pk fk] (apply utils/azipmap
                            (utils/kseq @rfk rpk))]
         (sfns/pred-= pk fk)))
```
can be written as

```clojure
(apply sfns/pred-and (map sfns/pred= @rfk rpk))
```

(assuming rfk & rpk are always sequences)

You're right about the foreign keys in many-to-many, I didn't think about it. However, are fk, lfk and rdk really necessary? What if we would accept a vector (or any seq) of columns instead?

Re with-batch: I guess in sql it would be possible to use tuples for composite keys in IN (e.g. where (id, i2) IN ((1,1), (2,2), ...).) Not sure how difficult that would be to do with Korma. Let's leave this out now and fix the rest of the PR first.

k13gomez commented:
I refactored pk, fk, lfk, rfk to always create a sequence of keys, and corrected the test cases accordingly.

The compatibility issue with zipmap is because using zipmap in 1.3 vs in 1.6 results different order of kv pairs, which results in slightly different (though still valid) sql. However, you're right about zipmap/azipmap being too complicated and I implemented your suggestion now that pk, fk, lfk, rfk are always sequences.

Regarding fk, lfk and rfk, they're not absolutely necessary nor mandatory, but I think they're a nice way to write these relationships to avoid having to wrap a fk/lfk/rfk value in '( ). The following are equivalent under current implementation:
```clojure
(many-to-many mtm2 :mtm2_mtm1 (lfk :foo)
                                                         (rfk :bar))
(many-to-many mtm2 :mtm2_mtm1 {:lfk '(:foo)
                                                          :rfk '(:bar)})
```
```clojure
(has-one mtm2 (fk :foo))
(has-one mtm2 {:fk '(:foo)})
```

immoh commented:
I added couple of inline comments, otherwise it's starting to look good. Fk helper macros are ok, I would personally use vectors `[:foo :bar:]` instead of list `'(:foo :bar)` but his way users can use whatever they want. 

Can you also check that all docstrings are up-to-date?

Thanks.

k13gomez commented:
Take a look at the latest changes when you have a chance, I think I addressed everything; also, vectors do work just fine for defining fk/lfk/rfk functions, I verified it.

immoh commented:
It's overkill to test and document that vectors and lists behave in the same way here, it is a property of Clojure and we can leave it out. In my opinion we should only use helper functions (fk, lfk & rfk) in docstring and tests.

Otherwise it looks really good! When you're done, please squash the commits.

k13gomez commented:
Alright, done and squashed.

immoh commented:
Thanks!

immoh commented:
Can I ask you to also prepare a pull request for https://github.com/korma/sqlkorma with changes for composite keys? Thanks.

k13gomez commented:
Sure thing, I'll take care of it as soon as I have a chance.

On Tue, Jun 16, 2015 at 6:19 AM, Immo Heikkinen <notifications@github.com>
wrote:

> Can I ask you to also prepare a pull request for
> https://github.com/korma/sqlkorma with changes for composite keys? Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/korma/Korma/pull/303#issuecomment-112375030>.
>



-- 
Jose Gomez
Software Engineer

k13gomez commented:
Take a look at PR [#13](https://github.com/korma/sqlkorma/pull/13) for https://github.com/korma/sqlkorma
